### PR TITLE
New CharacterNameManager

### PR DESCRIPTION
**Pull request by TomMalbran**
I moved the loading of the character names into a new CharacterNameManager, from where you can get new names.

When loading names they are randomized and added to an unused names queue. Each time you ask for a new name, it retrieves the first from the list. This makes it so that you don't get repeated names, until all the names are used, and we start using all the used names again.

This also makes it possible for a mod to add its own names.

### CODE REVIEW

File: Assets/Scripts/Models/CharacterNameManager.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    /// <summary>
+    /// Load new names to use for Characters and add them in a random way.
+    /// </summary>
+    /// <param name="nameStrings">An array of name strings.</param>
+    public static void LoadNames(string[] nameStrings)
+    {
+        // Randomize the given strings
+        for (int i = 0; i < nameStrings.Length; i++)
```
Comment (sboigelot) [inline Assets/Scripts/Models/CharacterNameManager.cs]: This is very verbose and low level coding.

What about nameStrings.OrderBy(Unity.Random.Range(0f, 1f)) to replace this whole block of code?

File: Assets/Scripts/Models/CharacterNameManager.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        foreach (string name in nameStrings)
+        {
+            unusedNames.Enqueue(name);
+        }
+    }
+
+    /// <summary>
+    /// Gets a name which wasn't used before.
+    /// </summary>
+    /// <returns>Psossibly an unused name.</returns>
```
Comment (mikejbrown) [inline Assets/Scripts/Models/CharacterNameManager.cs]: Possibly -- also this conflicts a bit with the summary. Could even drop the returns tag... I would suggest something like
```c#
/// <summary>
/// Returns a randomly chosen name, prioritizing names which have not been used yet.
/// </summary>
```

File: Assets/Scripts/Models/CharacterNameManager.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+// This program comes with ABSOLUTELY NO WARRANTY; This is free software,
+// and you are welcome to redistribute it under certain conditions; See
+// file LICENSE, which is part of this source code package, for details.
+// ====================================================
+#endregion
+using System.Collections.Generic;
+using UnityEngine;
+
+/// <summary>
+/// Character name manager that holds all the possible names and tryes to give a random and unused name everytime is requested.
```
Comment (mikejbrown) [inline Assets/Scripts/Models/CharacterNameManager.cs]: tries

File: Assets/Scripts/Models/CharacterNameManager.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    /// Initializes a new instance of the <see cref="CharacterNameManager"/> class.
+    /// </summary>
+    public CharacterNameManager()
+    {
+        unusedNames = new Queue<string>();
+        usedNames = new Queue<string>();
+    }
+
+    /// <summary>
+    /// Load new names to use for Characters and add them in a random way.
```
Comment (mikejbrown) [inline Assets/Scripts/Models/CharacterNameManager.cs]: Don't need the second half if you change the doc for `GetNewName` as suggested. Just "Load new names to use for Characters." The fact that you do the randomization here and not in `GetNewName` is an invisible implementation detail which may change without changing the behavior in any important way.

File: Assets/Scripts/Controllers/ModsManager.cs [hunk 5]
```diff
@@ -47,27 +49,59 @@ public DirectoryInfo[] GetMods()
...
+            });
+    }
+
+    /// <summary>
+    /// Loads the given file from the given folder in the base and inside the mods and calls the Delegate.
+    /// </summary>
+    /// <param name="folderName">Folder name.</param>
+    /// <param name="fileName">File name.</param>
+    /// <param name="readText">Called to handle the text reading and actual loading.</param>
+    private void LoadTextFile(string folderName, string fileName, Action<string> readText)
```
Comment (mikejbrown) [inline Assets/Scripts/Controllers/ModsManager.cs]: Nice refactoring

File: Assets/Scripts/Controllers/ModsManager.cs [hunk 6]
```diff
@@ -47,27 +49,59 @@ public DirectoryInfo[] GetMods()
...
+            fileName,
+            (filePath) =>
+            {
+                string[] lines = File.ReadAllLines(filePath);
+                CharacterNameManager.LoadNames(lines);
+            });
+    }
+
+    /// <summary>
+    /// Loads the given file from the given folder in the base and inside the mods and calls the Delegate.
```
Comment (mikejbrown) [inline Assets/Scripts/Controllers/ModsManager.cs]: Would say something like "...and calls the delegate with the text of each file."

File: Assets/Scripts/Models/CharacterNameManager.cs (hunk 1 above)
Comment (TomMalbran) [inline Assets/Scripts/Models/CharacterNameManager.cs]: Sure. Had no idea I could do that.

**Comment by TomMalbran** [comment]
Fixed the comments and the randomizer. Thanks for the comments.

File: Assets/Scripts/Models/CharacterNameManager.cs (hunk 1 above)
Comment (TomMalbran) [inline Assets/Scripts/Models/CharacterNameManager.cs]: @sboigelot I tried your code and it didn't work. I searched for a way to do it, and the threads suggested to implement Shuffle, which is basically what I did before. So I reverted the change. If you know a way to make it work, let me know.

Comment (sboigelot) [inline Assets/Scripts/Models/CharacterNameManager.cs]: Check https://github.com/sboigelot/TwentyCardsToTheApocalypse/blob/master/Assets/Scripts/Models/Deck.cs

Lines

```
        public void Shuffle()
        {
            Cards = Cards.OrderBy(c => Random.Range(0f, 1f)).ToList();
        }
```

But it's on a List, should work on an array too

Comment (TomMalbran) [inline Assets/Scripts/Models/CharacterNameManager.cs]: Great. That worked.

**Comment by TomMalbran** [comment]
Now the Randomizer works.

**Pull request merged**